tests: Equality on addresses, slices, cells.

jeshecdom opened the pull request:
## Issue

Towards #539.

## Checklist

- [ ] I have updated CHANGELOG.md
- [X] I have run all the tests locally and no test failure was reported
- [X] I have run the linter, formatter and spellchecker
- [X] I did not do unrelated and/or undiscussed refactorings

jeshecdom committed:
Added semantics tests for equality on addresses, slices, cells.

anton-trunov commented on src/test/e2e-emulated/contracts/semantics.tact:
```diff
@@ -1886,4 +1911,216 @@ contract SemanticsTester {
+        // Addresses are immutable.
+
+        let addr1 = myAddress();
+        let addr2 = contractAddress(initOf SemanticsTester());
```
Let's also add this equality to the docs, if it's not there yet

novusnota replied:
We can add something to `initOf()` for the time being. Later, this moment can be expanded upon in #925 and #1200

jeshecdom replied:
I am not quite sure of what I need to add to the docs. Do you mean the equivalence between obtaining the contract address with `myAddress()` and indirectly through `initOf`?

novusnota replied:
@jeshecdom yes, something like that. If you don't mind, I'll add it myself to this PR in a couple of hours :)

UPD: Actually, this will happen within this hour — I'm already at the laptop
UPD2: By the way, the `contractAddress(initOf ...())` approach is about 2.5k gas more expensive than just `myAddress()`, which only uses a single instruction.

novusnota replied:
Done! Also improved the similar descriptions on the operators page

anton-trunov commented on src/test/e2e-emulated/contracts/semantics.tact:
```diff
@@ -1886,4 +1911,216 @@ contract SemanticsTester {
+
+    /*************** Cells ********************/
+
+    get fun testCellEquality(): Bool {
```
Let's add tests for cells obtained in a different way, not just from structs

anton-trunov reviewed (changes requested):
Cool stuff, just some minor comments

novusnota committed:
feat(docs): equality of `myAddress()` and `contractAddress(initOf CurrentContract())`

And small updates of similar descriptions on the Operators page of the
Book

jeshecdom committed:
Added tests of equality of slices with references to cells.
Added tests of equality of cells manually built using builders (with cells having references and without).

jeshecdom committed:
Merge branch 'semantics-test' of https://github.com/tact-lang/tact into semantics-test

jeshecdom committed:
Merge branch 'main' into semantics-test
